### PR TITLE
Moving user tokens into firestore

### PR DESCRIPTION
**Pull request by gauntface**
This moves tokens into firestore rather than keeping them in memory on the server.

The original desire from everyone was to have tokens nested under a user, for example:

> users > ${username} > tokens > ${token} > ${tokenDetails}

This presents a problem when looking up information when we only have the token. Firestore only supports simply search of document fields on a collection (i.e. search for a user or search for token details).

I ended up using:

> tokens > ${token} > ${token Details}

This can be search for as `Find tokens where the username == 'gauntface'` allowing information look up by a token or a users login.

### CODE REVIEW

File: src/server/models/userModel.ts [hunk 1]
```diff
@@ -58,13 +65,17 @@ class UserModel {
...
+
     const details = {
       username: loginResult.data.viewer.login,
-      token,
       scopes,
+      token,
     };
-
-    this.cachedLogins[token] = details;
+
```
Comment (samuelli) [inline src/server/models/userModel.ts]: nit: trailing whitespace - did formatter run?

File: src/server/models/userModel.ts [hunk 2]
```diff
@@ -11,21 +12,21 @@ interface LoginDetails {
...
+      return null;
+    }
+    
+    const loginDetails = query.docs[0].data();
+    if (!loginDetails) {
+      return null;
     }
-    return null;
+    
+    return (loginDetails as LoginDetails);
```
Comment (aomarks) [inline src/server/models/userModel.ts]: no need for parens

File: src/server/models/userModel.ts [hunk 3]
```diff
@@ -11,21 +12,21 @@ interface LoginDetails {
 }
 
 class UserModel {
-  private cachedLogins: {[id: string]: LoginDetails};
+  async getLoginDetails(login: string): Promise<LoginDetails|null> {
+    const tokensCollection = await firestore()
```
Comment (aomarks) [inline src/server/models/userModel.ts]: Unclear what kind of tokens these are. Can you add "github" to the naming in the db and vars here?

File: src/server/models/userModel.ts [hunk 4]
```diff
@@ -11,21 +12,21 @@ interface LoginDetails {
...
-    this.cachedLogins = {};
-  }
-
-  async getLoginDetails(username: string) {
-    for (const token of Object.keys(this.cachedLogins)) {
-      const details = this.cachedLogins[token];
-      if (details.username === username) {
-        return details;
-      }
+    const query = await tokensCollection.where('username', '==', login).get();
```
Comment (aomarks) [inline src/server/models/userModel.ts]: FTR every document field is automatically indexed, so this should perform well: https://firebase.google.com/docs/firestore/query-data/indexing

File: src/server/models/userModel.ts [hunk 5]
```diff
@@ -11,21 +12,21 @@ interface LoginDetails {
 }
 
 class UserModel {
-  private cachedLogins: {[id: string]: LoginDetails};
+  async getLoginDetails(login: string): Promise<LoginDetails|null> {
```
Comment (aomarks) [inline src/server/models/userModel.ts]: It was nice in your subscriptions PR that you had a block comment describing the database layout. Can you do the same here somewhere?

**Comment by aomarks** [comment]
> This presents a problem when looking up information when we only have the token. Firestore only supports simply search of document fields on a collection (i.e. search for a user or search for token details).

What's the use case where we only have the token?

**Comment by gauntface** [comment]
@samuelli and @aomarks should be good for review.

The two scenarios we need information are:
- Browser makes a request and the only information we know if the Github Token via the request cookie.
- Github WebHook that provides a Github username and we need to get a token related to it.

**Comment by samuelli** [comment]
> Browser makes a request and the only information we know if the Github Token via the request cookie.

FWIW, this is a self imposed requirement. We could also store their GitHub login in the cookie alongside the token. However, we cannot trust the login provided in the cookie, so we would need to verify it against the tokens we have.

**Comment by gauntface** [comment]
I think it's a sane pattern to follow for the reason you've pointed out and if the reason I've structure the data this way.

**Pull request merged**